consensus/polybft/fsm.go

```diff
@@ -0,0 +1,558 @@
+
+	// uptimeCounter holds info about number of times validators sealed a block (only present if isEndOfEpoch is true)
+	uptimeCounter *UptimeCounter
+
+	// isEndOfEpoch indicates if epoch reached its end
+	isEndOfEpoch bool
+
+	// isEndOfSprint indicates if sprint reached its end
+	isEndOfSprint bool
+
+	// current epoch
+	epoch uint64
+
+	// proposerCommitmentToRegister is a commitment that is registered via state transaction by proposer
+	proposerCommitmentToRegister *CommitmentMessageSigned
+
+	// commitmentToSaveOnRegister is a commitment that is verified on register and needs to be saved in db
+	commitmentToSaveOnRegister *CommitmentMessageSigned
+
+	// bundleProofs is an array of bundles to be executed on end of sprint
+	bundleProofs []*BundleProof
+
+	// commitmentsToVerifyBundles is an array of commitment messages that were not executed yet,
+	// but are used to verify any bundles if they are included in state transactions
+	commitmentsToVerifyBundles []*CommitmentMessageSigned
+
+	// stateSyncExecutionIndex is the next state sync execution index in smart contract
+	stateSyncExecutionIndex uint64
+
+	logger hcf.Logger // The logger object
+}
+
+func (f *fsm) Init(info *pbft.RoundInfo) {
+	f.blockBuilder.Reset()
+	f.commitmentToSaveOnRegister = nil
+}
+
+// BuildProposal builds a proposal for the current round (used if proposer)
+func (f *fsm) BuildProposal() (*pbft.Proposal, error) {
+	parent := f.parent
+
+	extraParent, err := GetIbftExtra(parent.ExtraData)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: we will need to revisit once slashing is implemented
+	extra := &Extra{Parent: extraParent.Committed}
+
+	if f.isEndOfEpoch {
+		// TODO: Nemanja - no transactions for now
+		// tx, err := f.createValidatorsUptimeTx()
+		// if err != nil {
+		// 	return nil, err
+		// }
+		// if err := f.blockBuilder.CommitTransaction(tx); err != nil {
+		// 	return nil, fmt.Errorf("failed to commit validators uptime transaction: %v", err)
+		// }
+		validatorsDelta, err := f.getValidatorSetDelta(f.blockBuilder.GetState())
+		if err != nil {
+			return nil, err
+		}
+
+		extra.Validators = validatorsDelta
+		f.logger.Trace("[FSM Build Proposal]", "Validators Delta", validatorsDelta)
+	}
+
+	if f.config.IsBridgeEnabled() {
+		// TODO: Nemanja - no transactions for now
+		// for _, tx := range f.stateTransactions() {
+		// 	if err := f.blockBuilder.CommitTransaction(tx); err != nil {
+		// 		return nil, fmt.Errorf("failed to commit state transaction. Error: %v", err)
+		// 	}
+		// }
+	}
+
+	// set the timestamp
+	parentTime := time.Unix(int64(parent.Timestamp), 0)
+	headerTime := parentTime.Add(f.config.BlockTime)
+
+	if headerTime.Before(time.Now()) {
+		headerTime = time.Now()
+	}
+
+	// fill the block with transactions
+	now := time.Now()
+
+	// TODO: Nemanja - skip transactions
+	// if err := f.blockBuilder.Fill(context.Background()); err != nil {
+	// 	return nil, err
+	// }
+
+	f.logger.Debug("Fill block", "time", time.Since(now))
```

Comment:
Are we ok with this additional load? Do we have lazy logger for that cases?

Reply:
This was added back when we were testing v3 client on testnet and addressing issues. Not sure do we still need such log.

> Do we have a lazy logger for that cases?
I'm not aware of it, but maybe @zivkovicmilos, @lazartravica or @ferranbt would know?

Reply:
It is Debug log level, during normal operation of a node it won't be called.

Reply:
Hovewer, it will call time.Since and do a syscall.

Reply:
We have changed that in another pr https://github.com/0xPolygon/polygon-edge/pull/789